test/DaemonExec.test.ts

```diff
@@ -0,0 +1,167 @@
+import * as Docker from 'dockerode'
+
+import { promises as fs, createReadStream, createWriteStream } from 'fs'
+
+import { resolve, join } from 'path'
+
+import { tmpdir } from 'os'
+import {
+	VolumeDefinition,
+	listImages,
+	listContainers,
+	ensureImageImport,
+	createContainer,
+	attachStreams,
+	stopContainer,
+	removeContainer,
+} from '../src/Daemon/DaemonExec'
+
+interface ExpectedStatus {
+	Error: unknown
+	StatusCode: number
+}
+
+const docker = new Docker()
+const testImage = 'ubuntu:18.04'
+const bogusImage = 'lskjflskjflsdkfjslkfjsdlfsdjflksdj'
+const inputFile = 'package.json'
+
+describe('DaemonExec', () => {
+	let container: Docker.Container
+	let tempDir: string
+	let outputFile: string
+	let errorFile: string
+
+	it('pulls a new image', async () => {
+		const initialInfos = await listImages(docker)
+		if (
+			initialInfos.some((info) =>
+				info.RepoTags.some((tag) => tag === testImage),
+			)
+		) {
+			// if image is already imported, remove it
+			const image = docker.getImage(testImage)
+			await image.remove()
+		}
+
+		await ensureImageImport(docker, testImage)
+		const newInfos = await listImages(docker)
+
+		expect(
+			newInfos.some((info) => info.RepoTags.some((tag) => tag === testImage)),
+		).toBeTruthy()
+	}, 60000)
+
+	it('pulls an existing image', async () => {
+		await ensureImageImport(docker, testImage)
+		const newInfos = await listImages(docker)
+
+		expect(
+			newInfos.some((info) => info.RepoTags.some((tag) => tag === testImage)),
+		).toBeTruthy()
+	})
+
+	it('fails pulling a nonexistent image', async () => {
+		await expect(ensureImageImport(docker, bogusImage)).rejects.toThrow()
+	})
+
+	it('creates a specified container', async () => {
+		container = await createContainer(docker, testImage, ['/bin/cat'], [])
+		expect(container.id).toBeDefined()
+	})
+
+	it('attaches streams to a specified container', async () => {
+		// inputs package.json to stdin, gets stdout and stderr in temp directory
+
+		tempDir = await fs.mkdtemp(join(tmpdir(), 'test'))
+
+		const stdinStream: NodeJS.ReadableStream = createReadStream(inputFile)
+
+		outputFile = resolve(tempDir, inputFile)
+		const stdoutStream: NodeJS.WritableStream = createWriteStream(outputFile)
+
+		errorFile = resolve(tempDir, 'error')
+		const stderrStream: NodeJS.WritableStream = createWriteStream(errorFile)
+
+		const stream = await attachStreams(
+			container,
+			stdinStream,
+			stdoutStream,
+			stderrStream,
+		)
+		expect(stream).toBeDefined()
```

Comment:
This assertion isn't needed right? If `attachStreams` fails, it should reject the promise rather than return `undefined` (so it'll never get to this line).

Reply:
There is a warning if there is no expect in the test. I would like to avoid warnings if possible.

Reply:
Fair enough. Could you add a comment explaining that's why you have the assertion?